cpp/include/cudf/search.hpp

```diff
@@ -0,0 +1,105 @@
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "cudf.h"
+
+
+namespace cudf {
+
+/**---------------------------------------------------------------------------*
+ * @brief Find smallest indices in sorted column where values should be 
+ *  inserted to maintain order
+ * 
+ * @note The @p column is required to be sorted in ascending order otherwise
+ *  the behaviour is undefined
+ * 
+ * @param column        Column to search in. Must be sorted in ascending order
+ * @param values        Values to search insert locations for
+ * @param nulls_as_largest If true, nulls are considered larger than valid
+ *  values, otherwise, nulls are considered smaller than valid values
+ * 
+ * @return gdf_column   Insertion points with the same shape as values
+ *---------------------------------------------------------------------------**/
+gdf_column lower_bound(gdf_column const& column,
+                       gdf_column const& values,
+                       bool ascending = true,
+                       bool nulls_as_largest = true);
+
+/**---------------------------------------------------------------------------*
+ * @brief Find largest indices in sorted column where values should be 
+ *  inserted to maintain order
+ * 
+ * @note The @p column is required to be sorted in ascending order otherwise
+ *  the behaviour is undefined
+ * 
+ * @param column        Column to search in. Must be sorted in ascending order
+ * @param values        Values to search insert locations for
+ * @param nulls_as_largest If true, nulls are considered larger than valid
+ *  values, otherwise, nulls are considered smaller than valid values
+ * 
+ * @return gdf_column   Insertion points with the same shape as values
+ *---------------------------------------------------------------------------**/
+gdf_column upper_bound(gdf_column const& column,
+                       gdf_column const& values,
+                       bool ascending = true,
+                       bool nulls_as_largest = true);
+
+/**---------------------------------------------------------------------------*
+ * @brief Find smallest indices in sorted table where values should be 
+ *  inserted to maintain order
+ * 
+ * @note The table @p t is required to be sorted in ascending order otherwise
+ *  the behaviour is undefined
+ * 
+ * @param t             Table to search in. Must be sorted in ascending order
+ * @param values        Values to search insert locations for
+ * @param desc_flags    Vector of whether corresponding column is sorted in
+ *  descending. If false, the column is considered sorted ascending
+ * @param nulls_as_largest If true, nulls are considered larger than valid
+ *  values, otherwise, nulls are considered smaller than valid values
+ * 
+ * @return gdf_column   Insertion points with the same shape as values
+ *---------------------------------------------------------------------------**/
+gdf_column lower_bound(table const& t,
```

Comment:
Is the single column API necessary if we have the `table` API?.

Reply:
I do plan to remove the single column implementation to use just the table one after I add google benchmarks and confirm that it's not too slow compared to what we have. There's more advantages to this too: The compilation time currently is a bit high and I hope it comes down after the consolidation/cleanup.

Reply:
Still plan to remove the single column API?

Reply:
OH, the API? I just removed the implementation. Ok brb.